src/OrbitGl/ViewportTest.cpp

```diff
@@ -0,0 +1,176 @@
+  EXPECT_EQ(viewport.GetHeight(), 2000);
+  EXPECT_EQ(viewport.GetWorldWidth(), 100.f);
+  EXPECT_EQ(viewport.GetWorldHeight(), 200.f);
+  EXPECT_EQ(viewport.GetWorldExtents(), Vec2(100, 200));
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(0.f, 0.f));
+
+  // Test: Changing the world width / height should not affect screen extents
+  viewport.SetWorldWidth(500.f);
+  EXPECT_TRUE(viewport.IsDirty());
+  viewport.ClearDirtyFlag();
+
+  viewport.SetWorldHeight(600.f);
+  EXPECT_TRUE(viewport.IsDirty());
+  viewport.ClearDirtyFlag();
+
+  EXPECT_EQ(viewport.GetWidth(), 1000);
+  EXPECT_EQ(viewport.GetHeight(), 2000);
+  EXPECT_EQ(viewport.GetWorldWidth(), 500.f);
+  EXPECT_EQ(viewport.GetWorldHeight(), 600.f);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(0.f, 0.f));
+  EXPECT_EQ(viewport.GetWorldExtents(), Vec2(100, 200));
+
+  // Test: Changing the world extent should not affect screen size and visible world size
+  viewport.SetWorldExtents(1000, 2000);
+  EXPECT_TRUE(viewport.IsDirty());
+  viewport.ClearDirtyFlag();
+
+  EXPECT_EQ(viewport.GetWidth(), 1000);
+  EXPECT_EQ(viewport.GetHeight(), 2000);
+  EXPECT_EQ(viewport.GetWorldWidth(), 500.f);
+  EXPECT_EQ(viewport.GetWorldHeight(), 600.f);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(0.f, 0.f));
+  EXPECT_EQ(viewport.GetWorldExtents(), Vec2(1000, 2000));
+
+  // Test: Scrolling (within correct range) should mark as dirty
+  viewport.Resize(100, 200);
+  viewport.ClearDirtyFlag();
+
+  viewport.SetWorldTopLeftX(100.f);
+  EXPECT_TRUE(viewport.IsDirty());
+  viewport.ClearDirtyFlag();
+
+  viewport.SetWorldTopLeftY(-50.f);
+  EXPECT_TRUE(viewport.IsDirty());
+  viewport.ClearDirtyFlag();
+
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(100.f, -50.f));
+
+  // Setting everything to the same values again should not mark the viewport as dirty
+  viewport.Resize(100, 200);
+  viewport.SetWorldWidth(500.f);
+  viewport.SetWorldHeight(600.f);
+  viewport.SetWorldTopLeftX(100.f);
+  viewport.SetWorldTopLeftY(-50.f);
+  viewport.SetWorldExtents(1000, 2000);
+  EXPECT_FALSE(viewport.IsDirty());
+}
+
+TEST(Viewport, ScrollingWithoutZoom) {
+  Viewport viewport(100, 200);
+  viewport.SetWorldExtents(200, 300);
+
+  viewport.SetWorldTopLeftX(100.f);
+  viewport.SetWorldTopLeftY(-50.f);
+
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(100.f, -50.f));
+
+  viewport.SetWorldTopLeftX(150.f);
+  viewport.SetWorldTopLeftY(-150.f);
+
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(100.f, -100.f));
+  EXPECT_EQ(viewport.IsDirty(), true);
+  viewport.ClearDirtyFlag();
+
+  // Scrolling further after clamping should not result in a dirty flag
+  viewport.SetWorldTopLeftX(200.f);
+  viewport.SetWorldTopLeftY(-200.f);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(100.f, -100.f));
+  EXPECT_EQ(viewport.IsDirty(), false);
+
+  // Resizing should clamp
+  viewport.SetWorldExtents(150, 250);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(50.f, -50.f));
+}
+
+TEST(Viewport, ScrollingWithZoom) {
```

Comment:
Could we also add a test where we zoom in the middle of the test and check if the point_new_position is where it should be?

Reply:
It should be covered by the coordinate conversion test. I've extended that one to include a zoomed and non-zoomed version of all tests.